### PR TITLE
[auto run retry tags] 4/n -  update auto run reexecution to use `will_retry` and `auto_retry_run_id` tags

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Updates the auto run reexecution daemon to look at the presence of a `will_retry` tag to determine if a run should be retried, rather than recomputing it from state. 

If a failed run does not have a `will_retry` tag, we recompute whether it should be retried to prevent against a deployment having different code and daemon versions

Associated internal pr:  https://github.com/dagster-io/internal/pull/12766

## How I Tested These Changes
new unit test for fallback behavior 

Manual testing:
Env: webserver/daemons running this branch, code server on an older version of dagster
```python
@dg.asset
def has_retry_policy(context):
    if context.run.parent_run_id is None:
        raise Exception("I failed")
    else:
        return 1 
```

Test 1
ran the asset - that run fails. Observe that the `will_retry` tag is not set. See the daemon pick up the run and launch a retry, indicating the fallback behavior worked. See that the `will_retry` tag has been added to the run, as has the `auto_retry_run_id` tag

Test 2
stop running the daemon. Run a few materializations of the asset, all fail. See that the `will_retry` tag is not added to any run. Start the daemon back up. The daemon processes the failed runs and retries them. See that the `will_retry` and `auto_retry_run_id` tags have been added to the original runs

Test 3 (also added a unit test for this case)
stop running the daemon. Materialize the asset, it fails. See that the `will_retry` tag is not added to the run. Stop the code server. Start the daemon back up. The idea here is to simulate the code server going down before the daemon processes the run. The daemon processes the run and sees it should retry, but cannot launch the retry because the code server is down. The daemon will not process this run again, so we need to make sure the `will_retry` tag is not True, since the run will never be retried. See that the `will_retry` tag on the run is `false`

## Changelog

The automatic run retry daemon has been updated so that there is a single source of truth for if a run will be retried and if the retry has been launched. Tags are now added to run at failure time indicating if the run will be retried by the automatic retry system. Once the automatic retry has been launched, the run ID of the retry is added to the original run.

### CODE REVIEW

**Comment by jamiedemaria** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26055?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26055** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26055?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26055?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26054** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26054?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25932** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25932?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26046** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26046?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py [hunk 1]
```diff
@@ -23,56 +24,19 @@
...
-            run.tags.get(RETRY_ON_ASSET_OR_OP_FAILURE_TAG),
-            default_value=default_retry_on_asset_or_op_failure,
-        )
-        if retry_number is not None:
+        if get_boolean_tag_value(run.tags.get(WILL_RETRY_TAG), default_value=False):
+            if not get_boolean_tag_value(run.tags.get(DID_RETRY_TAG), default_value=False):
+                _, run_group = check.not_none(instance.get_run_group(run.run_id))
+                yield (run, len(list(run_group)))
+        else:
+            retry_on_asset_or_op_failure = get_boolean_tag_value(
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:36]: this else statement needs to stay in sync with the corresponding part of `should_retry_run` so that we can log the engine event if necessary. thinking about way to refactor this so that we dont rely on remembering to keep the conditions the same

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py [hunk 2]
```diff
@@ -22,56 +24,24 @@
...
     for run in runs:
-        retry_number = get_retry_number(run)
-        retry_on_asset_or_op_failure = get_boolean_tag_value(
-            run.tags.get(RETRY_ON_ASSET_OR_OP_FAILURE_TAG),
-            default_value=default_retry_on_asset_or_op_failure,
-        )
-        if retry_number is not None:
+        if get_boolean_tag_value(run.tags.get(WILL_RETRY_TAG), default_value=False):
+            if not get_boolean_tag_value(run.tags.get(DID_RETRY_TAG), default_value=False):
+                _, run_group = check.not_none(instance.get_run_group(run.run_id))
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:33]: should we be calculating this 'retry number' in the same place that we are calculating whether or not it will retry? Otherwise it seems like they could get out of sync. that could even be the value of the tag - 'dagster/will_retry_number' or something?

**Review by gibsondan — DISMISSED** [review]

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py (hunk 2 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:46]: just in the UI afaik

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:46]: maybe we need to make sure it's in sync with https://github.com/dagster-io/dagster/blob/bd069b7f761fb422830a93d504afe4fefc523c81/python_modules/dagster/dagster/_core/execution/context/system.py#L807 

but i dont think we're currently doing anything to guarantee that. But i also don't know the details of how the known state is determined off the top of my head

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:46]: i think that's step retries though?

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py [hunk 3]
```diff
@@ -22,68 +24,54 @@
...
-            if any([run.run_id == run_.parent_run_id for run_ in run_group_list]):
-                return None
-            return len(run_group_list)
+def get_retry_number_if_should_retry(
+    run: DagsterRun,
+    instance: DagsterInstance,
+) -> Optional[int]:
+    """If the run should retry, returns the retry number. Otherwise, returns None."""
+    will_retry_tag_value = run.tags.get(WILL_RETRY_TAG)
+    if will_retry_tag_value is None:
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:33]: is there a test that would fail if we accidentally made this `if not will_retry_tag_value` instead of `if will_retry_tag_value is None` - I guess that wouldn't be the end of the world since in theory the value should be the same even if it needs to re-compute it here

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py [hunk 4]
```diff
@@ -206,3 +193,7 @@ def consume_new_runs_for_automatic_reexecution(
                 run,
                 engine_event_data=EngineEventData(error=error_info),
             )
+            # mark that this run will not be retried so that the tags reflect the state of the system
+            workspace_process_context.instance.add_run_tags(
+                run_id=run.run_id, new_tags={WILL_RETRY_TAG: "false"}
+            )
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:199]: this is a little weird / might be confusing, but as long as you can still reconstruct the sequence of events by looking at the event log, it is probably fine

**Review by gibsondan — DISMISSED** [review]
you identified one idempotency issue over DM where the daemon could fail at just the wrong time and emit a retry twice - sounds like we have a plan to avoid that by swapping the order that we apply things during the launch and checking to see if the retried run in the tag was already created

The really tricky thing is the case where the run was created and then the daemon crashed before it could be submitted - in that case we probably would want to re-submit that run again

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py (hunk 4 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:199]: yeah i agree that it's weird, especially since it's really the daemon cursoring that makes updating this tag necessary

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py (hunk 3 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:33]: The values `will_retry_tag_value` could be are the string `"true"`, `"false"`, or `None` so i think if the behavior were changed to `if not will_retry_tag_value` the behavior would stay the same, right?

Comment (gibsondan) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:33]: that is a good point

**Comment by jamiedemaria** [comment]
> The really tricky thing is the case where the run was created and then the daemon crashed before it could be submitted - in that case we probably would want to re-submit that run again


is there a status that a created but not submitted run would have? or some other way to tell if we're in this case?

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py [hunk 5]
```diff
@@ -152,8 +139,27 @@ def retry_run(
...
+    _, run_group = check.not_none(instance.get_run_group(failed_run.run_id))
+    run_group_list = list(run_group)
+
+    # it is possible for the daemon to die between creating the run and submitting it. We account for this
+    # possibility by checking if the a run already exists in the run group with the parent run id of the
+    # failed run and resubmit it if necessary.
+    existing_retried_run = get_retried_run_if_exists(failed_run, run_group_list)
+    if existing_retried_run is not None:
+        # ensure the failed_run has the AUTO_RETRY_RUN_ID_TAG set
+        if failed_run.tags.get(AUTO_RETRY_RUN_ID_TAG) is None:
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:151]: [1]

There is an issue with adding the auto-retry tag here since this would also capture manual retries that are launched before the daemon processes this run for a manual retry

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py [hunk 6]
```diff
@@ -643,6 +806,8 @@ def test_consume_new_runs_for_automatic_reexecution_with_manual_retry(instance,
     # it is in the run group of the original run, we will see that the run group contains a run with
     # parent_run_id as the original run. Therefore the original run will not be retried
     assert len(instance.run_coordinator.queue()) == 0
+    run = instance.get_run_by_id(run.run_id)
+    assert run.tags.get(AUTO_RETRY_RUN_ID_TAG) == manual_retry.run_id
```
Comment (jamiedemaria) [inline integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:810]: example of [1] - this seems bad and unwanted

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py (hunk 5 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:151]: example:
run A fails
manual retry of run A `A_m` is launched
retry daemon processes run A 

with the current daemon behavior, no new run would be created, basically considering the manual retry as the retry that would have been launched

with this PR, we would maintain that behavior, but we still want a way to mark on run A that the retry has been launched (ie that it's not waiting for a retry to get launched). But setting the id of the manual run as `auto_retry_run_id` feels wrong

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py [hunk 7]
```diff
@@ -48,6 +48,21 @@ def create_run(instance, **kwargs):
...
+    status of each run to STARTED, since the auto-reexecution logic checks the status of the runs.
+    """
+    for run in queue:
+        updated_run = instance.get_run_by_id(run.run_id)
+        if updated_run.status == DagsterRunStatus.NOT_STARTED:
+            launch_started_event = DagsterEvent(
+                event_type_value=DagsterEventType.PIPELINE_STARTING.value,
+                job_name=run.job_name,
+            )
+            instance.report_dagster_event(launch_started_event, run_id=run.run_id)
```
Comment (gibsondan) [inline integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:63]: moving it to QUEUED rather than STARTING would be more like the queued run coordinator that it will actually be using in most cases, would that change anything to make that change?

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py [hunk 8]
```diff
@@ -152,8 +139,27 @@ def retry_run(
...
     )
 
-    strategy = get_reexecution_strategy(failed_run, instance) or DEFAULT_REEXECUTION_POLICY
+    _, run_group = check.not_none(instance.get_run_group(failed_run.run_id))
+    run_group_list = list(run_group)
+
+    # it is possible for the daemon to die between creating the run and submitting it. We account for this
+    # possibility by checking if the a run already exists in the run group with the parent run id of the
+    # failed run and resubmit it if necessary.
+    existing_retried_run = get_retried_run_if_exists(failed_run, run_group_list)
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:148]: should we move the AUTO_RETRY_RUN_ID_TAG tag check into this method and make get_automatically_retried_run_if_exists? It seems like the current impl would also return a manually retried run and short-circuit in the (admittedly unlikely) event that one was created before this auto-retry run kicked off, but i think we can pretty much just ignore manually retried runs for the purpose of this "check if the daemon died" check right?

Like i'm imagining:
- for whatever reason the auto retry daemon is delayed a few minutes
- user gets impatient, launches a manually retried run
- this check will now detect the manually retried run and not automatically retry right? Or if the manually retried run happens to be in a NOT_STARTED state, it will submit it again (even though it might be already being submitted by some other process)

**Review by gibsondan — CHANGES_REQUESTED** [review]
ok this broadly makes sense to me but one last edge case to consider with the new get_retried_run_if_exists codepath

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py (hunk 7 above)
Comment (jamiedemaria) [reply integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:63]: nope, having it be QUEUED would be fine. It just needs to be a status other than NOT_STARTED so that the daemon doesn't hit the "the daemon died between creating the run and submitting the run path"

File: integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py [hunk 9]
```diff
@@ -48,6 +48,21 @@ def create_run(instance, **kwargs):
         )
 
 
+def process_runs_in_queue(instance, queue: Sequence[DagsterRun]):
+    """The MockedRunCoordinator doesn't update the status of a run when submit_run is called. All other
```
[the remark this answers is not in the document]
Comment (jamiedemaria) [reply integration_tests/test_suites/daemon-test-suite/auto_run_reexecution_tests/test_auto_run_reexecution.py:52]: yeah i avoided doing any changes to the MockedRC bc i didn't want to deal with getting an instance available...

File: python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py (hunk 8 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/_daemon/auto_run_reexecution/auto_run_reexecution.py:148]: > for whatever reason the auto retry daemon is delayed a few minutes
user gets impatient, launches a manually retried run
this check will now detect the manually retried run and not automatically retry right? Or if the manually retried run happens to be in a NOT_STARTED state, it will submit it again (even though it might be already being submitted by some other process)


yeah that is the behavior that will happen both in master and in this PR as it stands right now 

If we change `get_retried_run_if_exists` to `get_automatically_retried_run_if_exists` are you thinking the impl would be like
```python
def get_automaticall_retried_run_if_exists(
    run: DagsterRun, instance,
) -> Optional[DagsterRun]:
    if run.tags.get(AUTO_RETRY_RUN_ID_TAG, None) is not None:
         return instance.get_run_by_id(run.tags[AUTO_RETRY_RUN_ID]
```

because then i think we still have to deal with the edge case where the daemon fails between creating the run and adding the tag to the original run

so maybe 
```python
def get_automaticall_retried_run_if_exists(
    run: DagsterRun, run_group: Sequence[DagsterRun], instance
) -> Optional[DagsterRun]:
    if run.tags.get(AUTO_RETRY_RUN_ID_TAG, None) is not None:
         return instance.get_run_by_id(run.tags[AUTO_RETRY_RUN_ID]
   child_run = next(
        (retried_run for retried_run in run_group if run.run_id == retried_run.parent_run_id), None
    )
   if child_run.status == DagsterRunStatus.NOT_STARTED:
      # if the run is in NOT_STARTED status then assume the daemon failed between creating the run and adding the tag to the parent run
      instance.add_run_tags(run.run_id, {AUTO_RETRY_RUN_ID_TAG: child_run.run_id}
     return child run
```

Then there's still a chance that the manual retry is timed just right to be NOT_STARTED when this fn runs, but that seems so unlikely

**Review by gibsondan — APPROVED** [review]
